### PR TITLE
feat: Implement getLocationCredentials handler & integrate with adapter

### PR DESCRIPTION
**Pull request by jimblanc**
#### Description of changes

Implements the `getLocationCredentials` handler as well as a wrapping API for `getDataAccess`. Integrates new handler with the managed auth adapter. Unit tests will be added shortly.

#### Description of how you validated changes
Local function testing in sample app. Tested prefixes, buckets, and object access with the managed auth adapter.

User-agent: `aws-amplify/6.3.6 storage/8 framework/1`

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/types/options.ts [hunk 1]
```diff
@@ -16,6 +16,16 @@ import {
...
+/**
+ * @internal
+ */
+export type Privilege = 'Default' | 'Minimal';
+
+/**
+ * @internal
+ */
+export type PrefixType = 'Object';
+
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/types/options.ts:28]: This can be moved to `storageBrowser/apis/types.ts` for now?

Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/types/options.ts:28]: Nit: moving these types to `storageBrowser/types.ts`? We can reorganize the `types.ts` into a folder later.

**Review by AllanZhengYP — APPROVED** [review]
Looks good!

**Review by AllanZhengYP — DISMISSED** [review]
Looks good!

File: packages/storage/src/storageBrowser/apis/getDataAccess.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import { GetDataAccessInput, GetDataAccessOutput } from './types';
+import { DEFAULT_CRED_TTL } from './constants';
+
+export const getDataAccess = async (
+	input: GetDataAccessInput,
+): Promise<GetDataAccessOutput> => {
+	const targetType = input.scope.endsWith('*') ? undefined : 'Object';
+	const { credentials } = await input.credentialsProvider();
+
+	const result = await getDataAccessClient(
```
Comment (AllanZhengYP) [inline packages/storage/src/storageBrowser/apis/getDataAccess.ts:22]: Add the info logger message here?

Comment (jimblanc) [reply packages/storage/src/storageBrowser/apis/getDataAccess.ts:22]: Used debug to align with other storage API logging

**Review by AllanZhengYP — APPROVED** [review]

File: packages/aws-amplify/package.json
Comment (jimblanc) [inline packages/aws-amplify/package.json:1]: Changes due to new user agent key (few bytes)

File: packages/storage/src/storageBrowser/apis/constants.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+export const DEFAULT_CRED_TTL = 15 * 60; // 15 minutes
```
Comment (ashika112) [inline packages/storage/src/storageBrowser/apis/constants.ts:4]: is this the recommended?

File: packages/storage/src/storageBrowser/apis/getDataAccess.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import { StorageError } from '../../errors/StorageError';
+import { logger } from '../../utils';
+
+import { GetDataAccessInput, GetDataAccessOutput } from './types';
+import { DEFAULT_CRED_TTL } from './constants';
+
+export const getDataAccess = async (
+	input: GetDataAccessInput,
+): Promise<GetDataAccessOutput> => {
+	const targetType = input.scope.endsWith('*') ? undefined : 'Object';
```
Comment (ashika112) [inline packages/storage/src/storageBrowser/apis/getDataAccess.ts:20]: curious: what does this '*' signify? i see Object is the only accepted value in s3 API

**Review by ashika112 — APPROVED** [review]

File: packages/storage/src/storageBrowser/apis/getDataAccess.ts (hunk 4 above)
Comment (AllanZhengYP) [reply packages/storage/src/storageBrowser/apis/getDataAccess.ts:20]: It's to signify a subPrefix of an AccessGrant. Se https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-grants-grant.html#subprefix, which is also the omission value of `TargetType`

File: packages/storage/src/storageBrowser/apis/constants.ts (hunk 3 above)
Comment (AllanZhengYP) [reply packages/storage/src/storageBrowser/apis/constants.ts:4]: @ashika112 For security posture, we use the lowest allowed TTL

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**